src/malloc_io.c

```diff
@@ -664,6 +664,33 @@ malloc_printf(const char *format, ...) {
 	va_end(ap);
 }
 
+void
+buf_writer_flush(buf_writer_arg_t *arg) {
+	arg->buf[arg->buf_end] = '\0';
+	if (arg->write_cb == NULL) {
+		arg->write_cb = je_malloc_message != NULL ?
+		    je_malloc_message : wrtmessage;
+	}
+	arg->write_cb(arg->cbopaque, arg->buf);
+	arg->buf_end = 0;
+}
+
+void
+buffered_write_cb(void *buf_writer_arg, const char *s) {
+	buf_writer_arg_t *arg = (buf_writer_arg_t *)buf_writer_arg;
+	size_t i, n, slen;
+	for (i = 0, slen = strlen(s); i < slen; i += n) {
+		if (arg->buf_end == arg->buf_size) {
```

Comment:
Move this to the end of the loop?

Reply:
The logic might be slightly different. Having the test and the flush upfront makes flushing lazy: if the input exactly fits the entire buffer, the flush won't take place until the next call of `buffered_write_cb()` or `buf_writer_flush()`. (My unit tests are particularly testing for this behavior.) The advantage is that the final `buf_writer_flush()` call would never unnecessarily call the real `write_cb()` one more time with an empty buffer. (Of course I can guard against an empty buffer in the `buf_writer_flush()` function, but that feels a bit redundant.)

Reply:
Ah I see. Didn't think of the final `buf_writer_flush()` call. Makes sense.